File: src/main/backends/darwin/request.m
//
//  request
//  libreq
//
//  Created by Rhys Adams on 11/03/2022.
//  Copyright 2022. All rights reserved.
//

#include <Foundation/Foundation.h>

#include "../../../../lib/b64.c/b64.h"

#include <request.h>

Res_t* req_get(const char* url, int use_proxy, const char* proxy, Headers_t* headers){
    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];

    NSMutableURLRequest* mutreq = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:[NSString stringWithUTF8String:url]]];

    if(headers != NULL) {
        unsigned int i;
        for (i=0; i < headers->length; i++) {
            [mutreq addValue:[NSString stringWithUTF8String:headers->values[i]] forHTTPHeaderField:[NSString stringWithUTF8String:headers->keys[i]]];
        }
    }

    NSURLRequest* req = [mutreq copy];

    NSURLResponse* res = nil;
	
	NSError* err = nil;
	
	NSData* data = [NSURLConnection sendSynchronousRequest:req returningResponse:&res error:&err];
	
	Res_t* out_var = malloc(sizeof(struct Res));
	out_var->length = 0;
	
	if(err == nil){
		out_var->err = 0;
		
		out_var->length = [data length];
		
		const void* bytes = [data bytes];
		
		out_var->data = malloc(out_var->length);
		
		memcpy((char*) out_var->data, bytes, out_var->length);

        [pool release];
		
		return out_var;
	}
	
	out_var->err = 1;
	
	NSLog(@"%@", [err localizedDescription]);

    [pool release];
	
	return out_var;
}

Res_t* req_post_auth(const char* url, int use_proxy, const char* proxy, const char* data, Headers_t* headers, const char* username, const char* password){
    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];

    NSMutableURLRequest* mutreq = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:[NSString stringWithUTF8String:url]]];
    if(headers != NULL) {
        unsigned int i;
        for (i=0; i < headers->length; i++) {
            [mutreq addValue:[NSString stringWithUTF8String:headers->values[i]] forHTTPHeaderField:[NSString stringWithUTF8String:headers->keys[i]]];
        }
    }

    [mutreq setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];

    NSString *authStr = [NSString stringWithFormat:@"%@:%@", [NSString stringWithUTF8String:username], [NSString stringWithUTF8String:password]];
    NSData* authData = [authStr dataUsingEncoding:NSASCIIStringEncoding];
    const char* as = b64_encode([authStr UTF8String], [authStr length]);
    NSString *authValue = [NSString stringWithFormat:@"Basic %@", [NSString stringWithUTF8String: as]];
    [mutreq addValue:authValue forHTTPHeaderField:@"Authorization"];


    NSString* postData = [NSString stringWithUTF8String:data];
    [mutreq setHTTPMethod: @"POST"];
    [mutreq setHTTPBody: [postData dataUsingEncoding:NSUTF8StringEncoding]];

    NSURLRequest* req = [mutreq copy];

    NSURLResponse* res = nil;

    NSError* err = nil;

    NSData* resdata = [NSURLConnection sendSynchronousRequest:req returningResponse:&res error:&err];

    free(as);

    Res_t* out_var = malloc(sizeof(struct Res));
    out_var->length = 0;

    if(err == nil){
        out_var->err = 0;

        out_var->length = [resdata length];

        const void* bytes = [resdata bytes];

        out_var->data = malloc(out_var->length);

        memcpy((char*) out_var->data, bytes, out_var->length);

        [pool release];

        return out_var;
    }

    out_var->err = 1;

    NSLog(@"%@", [err localizedDescription]);

    [pool release];

    return out_var;
}

File: src/main/backends/qt/request.cpp
//
// Created by rhys on 11/08/22.
//

#include <request.h>

#include <QtNetwork/QNetworkAccessManager>
#include <QtNetwork/QNetworkRequest>
#include <QtNetwork/QNetworkReply>
#include <QEventLoop>
#include <QCoreApplication>
#include <iostream>

LIBRARY_API Res_t* req_get(const char* url, Headers_t* headers){
    int argc = 0;
    QCoreApplication a(argc, NULL);
    QNetworkAccessManager manager;

    QNetworkRequest request;
    request.setUrl(QString(url));

    if (headers != NULL) {
        for (int i = 0; i < headers->length; i++) {
            request.setRawHeader(headers->keys[i], headers->values[i]);
        }
    }

    QNetworkReply* reply = manager.get(request);
    while (!reply->isFinished()) {
        qApp->processEvents(QEventLoop::ExcludeUserInputEvents);
    }

    Res_t* out = (Res_t*) malloc(sizeof(Res_t));
    if(reply->error() == QNetworkReply::NoError){
        QByteArray data = reply->readAll();

        out->err = 0;
        out->length = data.length();
        out->data = strdup(data.toStdString().c_str());
    }else{
        out->err = 1;
        out->length = 0;
        out->data = NULL;
    }

    return out;
}

LIBRARY_API Res_t* req_post_auth(const char* url, const char* data, Headers_t* headers, const char* username, const char* password) {

}